daemon/logger/alilogs/alilogs_test.go

```diff
@@ -0,0 +1,316 @@
+		topic:            "demo_topic",
+		extraLogContents: extraContents,
+		client:           mockClient,
+		logGroup: &sls.LogGroup{
+			Topic: proto.String("demo_topic"),
+			Logs:  []*sls.Log{},
+		},
+		messages: make(chan *logger.Message, maximumLogsPerPut),
+	}
+
+	go stream.collectLogs()
+
+	var wg sync.WaitGroup
+	wg.Add(maximumLogsPerPut + 102)
+
+	for i := 0; i < maximumLogsPerPut+102; i++ {
+		go worker(stream, &wg)
+	}
+	wg.Wait()
+	time.Sleep(batchPublishFrequency)
+	stream.Close()
+	if mockClient.Topic != "demo_topic" {
+		t.Errorf("check topic fail, expect:%s, actual:%s", stream.topic, mockClient.Topic)
+	}
+	if len(mockClient.Logs) != maximumLogsPerPut+102 {
+		t.Errorf("check log number fail, expect:%v, actual:%v", maximumLogsPerPut+102, len(mockClient.Logs))
+	}
+
+}
+
+func worker(stream *logStream, wg *sync.WaitGroup) {
+	stream.Log(&logger.Message{
+		Line:      []byte("test log"),
+		Timestamp: time.Time{},
+	})
+	wg.Done()
+}
+
+func TestValidateOpt(t *testing.T) {
+	// endpointKey, projectKey, logstoreKey, labelsKey, envKey
+	opt := map[string]string{}
+	opt[endpointKey] = ""
+	err := ValidateLogOpt(opt)
+	if err == nil {
+		t.Errorf("check log opt fail: %v", opt)
+	}
+
+	opt[endpointKey] = "test-endpoint"
+	opt[projectKey] = ""
+	err = ValidateLogOpt(opt)
+	if err == nil {
+		t.Errorf("check log opt fail: %v", opt)
+	}
+
+	opt[projectKey] = "test-project"
+	opt[logstoreKey] = ""
+	err = ValidateLogOpt(opt)
+	if err == nil {
+		t.Errorf("check log opt fail: %v", opt)
+	}
+
+	opt[logstoreKey] = "test-logstore"
+	opt[labelsKey] = "attr1,attr2"
+	opt[envKey] = "e1=v1,e2=v2"
+	err = ValidateLogOpt(opt)
+	if err != nil {
+		t.Errorf("check log opt fail: %v", opt)
+	}
+
+	opt["error-key"] = "unsupported"
+	err = ValidateLogOpt(opt)
+	if err == nil {
+		t.Errorf("check log opt fail: %v", opt)
+	}
+}
+
+func TestCollectLogsSimple(t *testing.T) {
+	ec1 := &sls.LogContent{
+		Key:   proto.String("ex1"),
+		Value: proto.String("ex1 value"),
+	}
+	ec2 := &sls.LogContent{
+		Key:   proto.String("ex2"),
+		Value: proto.String("ex2 value"),
+	}
+	extraContents := []*sls.LogContent{ec1, ec2}
+	mockClient := NewMockClient()
+	mockClient.ErrType = 0
+	stream := &logStream{
+		endpoint:         "test-endpoint",
+		projectName:      "test-project",
+		logstoreName:     "test-logstore",
+		topic:            "demo_topic",
+		extraLogContents: extraContents,
+		client:           mockClient,
+		logGroup: &sls.LogGroup{
+			Topic: proto.String("demo_topic"),
+			Logs:  []*sls.Log{},
+		},
+		messages: make(chan *logger.Message, maximumLogsPerPut),
+	}
+
+	go stream.collectLogs()
+	stream.Log(&logger.Message{
+		Line:      []byte("this is test log 1"),
+		Timestamp: time.Time{},
+	})
+	stream.Log(&logger.Message{
+		Line:      []byte("this is test log 2"),
+		Timestamp: time.Time{},
+	})
+	stream.Log(&logger.Message{
+		Line:      []byte("this is test log 3"),
+		Timestamp: time.Time{},
+	})
+	time.Sleep(batchPublishFrequency)
```

Comment:
Can you check how this [test](https://github.com/docker/docker/blob/master/daemon/logger/awslogs/cloudwatchlogs_test.go#L387-L405) works by mocking the newTicker? So you don't need to sleep.